doc/sphinx/proof-engine/tactics.rst

```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+use the :tacn:`split` tactic to replace the current goal with subgoals :g:`A` and :g:`B`.
+For a goal :g:`A \/ B`, use :g:`apply or_introl` to replace the current goal with :g:`A`, or
+:g:`apply or_intror` to replace the current goal with :g:`B`.
+
+The other splitting patterns described below are not frequently used.
+
+These patterns can be used when the hypothesis is an equality:
+
+* :n:`->` - replaces the right-hand side of the hypothesis with the left-hand
+  side of the hypothesis in the conclusion of the goal; the hypothesis is
+  cleared; if the left-hand side of the hypothesis is a variable, it is
+  substituted everywhere in the context and the variable is removed.
+* :n:`<-` - similar to :n:`->`, but replaces the left-hand side of the hypothesis
+  with the right-hand side of the hypothesis.
+* :n:`[= @intropatterns ]` - If the product is over an equality type,
+  applies either :tacn:`injection` or :tacn:`discriminate`
+  instead of :tacn:`destruct`; if :tacn:`injection` is applicable, the patterns
+  :n:`{+, p}` are used on the hypotheses generated by :tacn:`injection`; if the
+  number of patterns is smaller than the number of hypotheses generated, the
+  pattern :n:`?` is used to complete the list.
+
+**need example for [= ... ]**
+
+* :n:`( @simple_intropattern , ... , @simple_intropattern )` - matches
+  a product over an inductive type :g:`I` with a single
+  constructor that itself has at least `n` arguments.  It performs a case
+  analysis over the hypothesis, as :n:`destruct` would, and applies the
+  patterns to the arguments of the constructor of :g:`I`.  Note that
+  that :n:`({+ p})` is equivalent to :n:`[{+ p}]`);
+* :n:`( @simple_intropattern & ... & @simple_intropattern )` - :n:`({+& p})`
+  is a shortcut for matching with
+  :n:`(p,( ... ,( ..., p ) ... ))`.  It expects the hypothesis to be a sequence of
+  right-associative binary inductive constructors such as :g:`conj` or
+  :g:`ex_intro`; for instance, a hypothesis with type
+  :g:`A /\(exists x, B /\ C /\ D)` can be introduced via pattern
+  :n:`(a & x & b & c & d)`;
+
+
+**need example for & ... &**
+
+**A /\\ (exists x, B /\\ C /\\ D) is not valid for (a & x & b & c & d)**
+
+* :n:`*` - introduces one or more quantified variables from the result
+* :n:`**` - introduces one or more quantified variables or hypotheses from the result until there are
+  no more quantified variables or implications (:g:`->`).  :g:`intros **` is equivalent
+  to :g:`intros`.
+
+* :n:`%term` - a pattern :n:`p{+ %term}` first applies :n:`{+ term}`
+  on the hypothesis to be introduced (as in :n:`apply {+, term}`) prior to the
+  application of the introduction pattern :n:`p`;
+
+  the on-the-fly application of lemmas: :n:`p{+ %term}` where :n:`p`
+  itself is not a pattern for on-the-fly application of lemmas (note:
+  syntax is in experimental stage)  **huh?**
+
+**need example, maybe simplify "intros [= -> %IHl]." from List.v?**
```

Comment:
Could this [example](https://github.com/tchajed/coq-tricks/blob/master/src/IntroPatterns.v) be of any help here?

Reply:
Thanks, I'll adapt that.  But without some explanation, understanding it takes a couple minutes of experimentation:
- `intros` would just add `H : S (length ys) = 1`
- `intros [=]` would reduce (right term?) to add `H0 : length ys = 0`
- `intros [=->%length_zero_iff_nil]` applies the resulting equality `l=[]`

I wish Coq had a way to show these multiple substeps for this example (and many, many others).  It's not so easy to understand what Coq is doing when there are multiple substeps.  I think this would be very helpful for learners.  But creating such a feature might be difficult but maybe feasible.  Certainly I'd have to learn far more about tactic internals.  (What do you think?)

It may be helpful to have a way to do diffs that comparing the result of applying 2 alternative tactics.  (Maybe).

Off topic: have you tried the diffs feature in master?  It will be in 8.9, but it looks like 2 needed bug fix PRs for it may not make it into 8.9 :-(.

Reply:
The `Info` command was an attempt to display such a decomposition but it is unfinished.

Reply:
> I wish Coq had a way to show these multiple substeps for this example (and many, many others).

I fully agree. Although, I'd say that stepping through a proof (with all those `;`) without dismantling it first would be a the most desirable feature.

> Off topic: have you tried the diffs feature in master?

Not yet. Does ProofGeneral support it?

Reply:
@anton-trunov No, diffs are not in PG yet.  I was hoping someone who's worked on PG would volunteer to do the small amount of work needed for diffs in PG once 8.9 is out.  Perhaps unrealistic, I'll see.

> I'd say that stepping through a proof (with all those ;) without dismantling it first would be a the most desirable feature.

Maybe it would not be hard to add an option that would treat ; as if it was "." for non-LTAC code.   I suspect it may be complex to robustly extend the concept to LTAC, which I don't know well.   Would an option that doesn't work in LTAC still be worthwhile?

Reply:
You cannot simply replace `;` with `.` and hope for the code to work though unless you add the proper goal selectors.